girder_worker_utils/decorators.py

```diff
@@ -16,13 +17,190 @@ class MissingInputException(Exception):
+            name, self._parameter_repr[self._signature.parameters[name].kind])
+            for name in self._signature.parameters]) + ")>"
+
+    def __getitem__(self, key):
+        return self._construct_argument(
+            self._get_class(self._signature.parameters[key]), key)
+
+    def _construct_argument(self, cls, name, **kwargs):
+        p = self._signature.parameters[name]
+        metadata = {}
+
+        if p.default != p.empty:
+            metadata['default'] = p.default
+        if p.annotation != p.empty:
+            # TODO: make sure annotation is a type and not just garbage
+            metadata['data_type'] = p.annotation
+
+        metadata.update(self._metadata.get(name, {}))
+
+        return cls(name, **metadata)
+
+    def _is_varargs(self, p):
+        return p.kind == Parameter.VAR_POSITIONAL
+
+    def _is_kwargs(self, p):
+        return p.kind == Parameter.VAR_KEYWORD
+
+    def _is_kwarg(self, p):
+        return p.kind == Parameter.KEYWORD_ONLY or (
+            p.kind == Parameter.POSITIONAL_OR_KEYWORD and p.default != p.empty)
+
+    def _is_posarg(self, p):
+        return p.kind == Parameter.POSITIONAL_ONLY or (
+            p.kind == Parameter.POSITIONAL_OR_KEYWORD and p.default == p.empty)
+
+    def _get_class(self, p):
+        if self._is_varargs(p):
+            return Varargs
+        elif self._is_kwargs(p):
+            return Kwargs
+        elif self._is_posarg(p):
+            return Arg
+        elif self._is_kwarg(p):
+            return KWArg
+        else:
+            raise RuntimeError("Could not determine parameter type!")
+
+
+    def set_metadata(self, name, key, value):
+        if name not in self._signature.parameters:
+            raise RuntimeError("{} is not a valid argument to this function!")
+
+        if name not in self._metadata:
+            self._metadata[name] = {}
+
+        self._metadata[name][key] = value
+
+    @property
+    def arguments(self):
+        # Only return arguments if we've declared them as paramaters
+        # This prevents us from returning things like 'self' of bound
+        # methods (e.g. celery tasks) etc.  This is a dubious design
+        # decision.
+        return [
+            self._construct_argument(
+                self._get_class(self._signature.parameters[name]), name)
+            for name in self._signature.parameters if name in self._metadata]
+
+    @property
+    def varargs(self):
+        for name in self._signature.parameters:
+            if name in self._metadata and \
+               self._is_varargs(self._signature.parameters[name]):
+                return self._construct_argument(Varargs, name)
+        return None
+
+    @property
+    def kwargs(self):
+        for name in self._signature.parameters:
+            if name in self._metadata and \
+               self._is_kwargs(self._signature.parameters[name]):
+                return self._construct_argument(Kwargs, name)
+        return None
```

Comment:
At the very least, we should rename these properties to indicate that they are a descriptor of the "target variable".  (If that makes any sense.)  The difference in handling these vs. the ordinary args/kwargs makes me lean towards not allowing them in decorated functions.  What advantage do they bring over a list or dict typed argument besides possibly an easier integration with click?

Reply:
renamed to PositionalArg, KeywordArg, VarsArg, KwargsArg